### PR TITLE
feat(direnv): add new direnv module

### LINKED ISSUE #4962: Add direnv module
## Feature Request
Similar to the `nix-shell` module, a `direnv` module would be great to show if direnv is active or an `.envrc` is present but not allowed.

### PR DESCRIPTION
**Pull request by Ajpantuso**
#### Description

Adds a new module `direnv` to display the status of the current rc file if one is present.

#### Motivation and Context

Closes #4962 

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

File: src/configs/starship_root.rs [hunk 1]
```diff
@@ -119,6 +119,7 @@ pub const PROMPT_ORDER: &[&str] = &[
     "os",
     "shell",
     "character",
+    "direnv",
```
Comment (mickimnet) [inline src/configs/starship_root.rs:122]: Is there a particular reason, why the `[direnv]` module is shown after the `[character]` module? I would be confused, if the `[character]` module wouldn't be the last?

File: docs/config/README.md [hunk 2]
```diff
@@ -1195,6 +1196,39 @@ truncation_length = 8
...
+| `detect_files` | `['.envrc']`                             | Which filenames should trigger this module. |
+
+### Variables
+
+| Variable | Example               | Description                             |
+| -------- | --------------------- | --------------------------------------- |
+| loaded   | `'loaded'`            | Whether the current rc file is loaded.  |
+| allowed  | `'denied'`            | Whether the current rc file is allowed. |
+| rc_path  | `'/home/test/.envrc'` | The current rc file path.               |
+| style\*  | `'red bold'`          | Mirrors the value of option `style`.    |
```
Comment (mickimnet) [inline docs/config/README.md:1219]: The variables in the documentation normally don't use quotation marks.

File: src/configs/starship_root.rs (hunk 1 above)
Comment (Ajpantuso) [reply src/configs/starship_root.rs]: Nope,just a hasty addition. Moved it before `env_var` since there's a natural relationship between the two modules.

File: docs/config/README.md (hunk 2 above)
Comment (Ajpantuso) [reply docs/config/README.md]: Fixed

**Comment by drupol** [comment]
What's the status of this PR ? It would be nice to move it forward, I'm looking for this feature.

**Comment by guettli** [comment]
BTW, you can help yourself with this (without an official direnv plugin):


# starship.toml
```
[custom.direnv]
command = '[[ $(direnv status) =~ "Found RC allowed false" ]] && echo "=========> missing: direnv allow <============"' # shows output of command
detect_files = ['.envrc'] # can specify filters but wildcards are not supported
```

File: src/modules/direnv.rs [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+            })
+            .map(|variable| match variable {
+                "rc_path" => Some(Ok(state.rc_path.to_string_lossy())),
+                "allowed" => Some(Ok(Cow::from(state.allowed.to_string()))),
+                "loaded" => state
+                    .loaded
+                    .then_some("loaded")
+                    .or(Some("not loaded"))
+                    .map(Cow::from)
+                    .map(Ok),
```
Comment (davidkna) [inline src/modules/direnv.rs:50]: I think it should be possible to customize the different `allowed`/`loaded`-strings (or symbols).

File: src/modules/direnv.rs [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+            return None;
+        }
+    });
+
+    Some(module)
+}
+
+fn log_error(error: impl Display) {
+    log::warn!("Error in module `direnv`:\n{}", error);
```
Comment (davidkna) [inline src/modules/direnv.rs:69]: We only tend to use this wording `module.set_segments`-part, so I would prefer removing this function, and to log errors/warnings directly (and generally without this prefix).

File: docs/config/README.md [hunk 5]
```diff
@@ -340,7 +341,7 @@ $status\
 $os\
 $container\
 $shell\
-$character"""
+$character """
```
Comment (davidkna) [inline docs/config/README.md:344]: Was this change intended?
```suggestion
$character"""
```

**Review by davidkna — CHANGES_REQUESTED** [review]

**Comment by Ajpantuso** [comment]
@davidkna I've applied your suggestions, but would like to keep	`rc_path` as an optional variable because I believe it will be a convenient debugging tool for non-developers.

Otherwise any input to make this module more consistent with other modules would be appreciated, thanks!

File: src/modules/direnv.rs [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            .set_extensions(&config.detect_extensions)
+            .set_files(&config.detect_files)
+            .set_folders(&config.detect_folders)
+            .is_match();
+
+    if !direnv_applies {
+        return None;
+    }
+
+    let direnv_status = &context.exec_cmd("direnv", &["status"])?.stdout;
```
[the remark this answers is not in the document]
Comment (Ajpantuso) [reply src/modules/direnv.rs:27]: Just to be clear this module would be disabled by default and even when enabled will only run if a `.envrc` file is present.

In the case that the module is enabled and a `.envrc` file is present running `direnv status` is the only way to determine if the file is allowed and loaded.

Can you help clarify what other information could be provided to this module to know if it is "needed" to run?

```rust
impl<'a> Default for DirenvConfig<'a> {
    fn default() -> Self {
        Self {
            format: "[$symbol$loaded/$allowed]($style) ",
            symbol: "direnv ",
            style: "bold orange",
            disabled: true,
            detect_extensions: vec![],
            detect_files: vec![".envrc"],
            detect_folders: vec![],
            allowed_msg: "allowed",
            denied_msg: "denied",
            loaded_msg: "loaded",
            unloaded_msg: "not loaded",
        }
    }
}
```

[the remark this answers is not in the document]
Comment (cathaysia) [reply src/modules/direnv.rs:27]: direnv will add some env vars when it loaded, such `$DIRENV_DIFF`, `DIRENV_DIR`, `DIRENVFILE` and `DIRENV_WATCH`. This should be helpful to do this.

[the remark this answers is not in the document]
Comment (cathaysia) [reply src/modules/direnv.rs:27]: Generally speaking I think IO operations and process operations should be avoided as much as possible, these are always slowly.

[the remark this answers is not in the document]
Comment (Ajpantuso) [reply src/modules/direnv.rs:27]: In that case this module could never provide information when a `.envrc` file is not allowed which was a request in #4962

[the remark this answers is not in the document]
Comment (cathaysia) [reply src/modules/direnv.rs:27]: if .direnv file existed but  has no env vars, I think it should be  consider as "disallowed".

In fact, I don't object to you running multiple processes, I just want to be able to run it when you need it.  For example, if I just want direnv to display a symbol when it is loaded, then the direnv process should not be running

[the remark this answers is not in the document]
Comment (cathaysia) [reply src/modules/direnv.rs:27]: Okay, so if I don’t plan to get the disallowed status, can I not run direnv?

[the remark this answers is not in the document]
Comment (cathaysia) [reply src/modules/direnv.rs:27]: Today I re-examined the environment variables of direnv and the implementation of p10k. you're right. It is not possible to obtain the disallowed status without running direnv.
p10k checks the environment variables, which means that the icon will be displayed even in the disallowed state. I agree with your changes. But is it possible if I don't intend to get the disallowed status (like p10k does).

[the remark this answers is not in the document]
Comment (Ajpantuso) [reply src/modules/direnv.rs:27]: I think that's a decision for the project maintainers, but I would say:

- Text parsing from the output of `direnv status` is already brittle let alone relying on environment variables used internally by `direnv` which have polluted the shell's environment.
- Using `direnv status` allows for extensibility in the future to collect and display more information via the prompt whereas the environment variables provide minimal information.
- Performance optimization without benchmarking/analysis is pre-optimization. To support your argument for deferring execution of `direnv status` I would like to see evidence that the latency incurred is statistically significant with regards to total execution time e.g. with execution of `direnv status` prompt display latency is 20% slower than without.
- This module is disabled by default so users are only affected by the additional overhead if they choose to be.

[the remark this answers is not in the document]
Comment (cathaysia) [reply src/modules/direnv.rs:27]: I don't think this is a pre-optimization, this is a foreseeable optimization point. The overhead of multi-processing is typically around two milliseconds on Linux and up to 8 milliseconds on Windows. For prompt, reaching 8 milliseconds for a single component is no longer a problem that can be simply ignored. For me, the latency I've tested that I can perceive is around 80 milliseconds. Every component in a starship should have performance considerations carefully considered.

And I add a benchmark for `direnv`, here are the results:

the benchmark command is `hyperfine --warmup 1000 'direnv status' -N`

`macos m2 chip`:

```console
Benchmark 1: direnv status
  Time (mean ± σ):       4.0 ms ±   0.4 ms    [User: 1.5 ms, System: 1.2 ms]
  Range (min … max):     3.2 ms …   6.2 ms    608 runs
```

[the remark this answers is not in the document]
Comment (Ajpantuso) [reply src/modules/direnv.rs:27]: It's pre-optimization in the sense that capabilities and design are being compromised for the sake of performance when performance has not yet been demonstrated to be an issue. Benchmarking executing `direnv` in isolation is also insufficient to understand the impact in a concurrent context especially when the primary function of the application is i/o bound i.e displaying text which is always slow compared to computation.

[the remark this answers is not in the document]
Comment (jankatins) [reply src/modules/direnv.rs:27]: Just FYI: starship executes [most things in parallel](https://github.com/starship/starship/blob/ba9f209a5fa739e6d183d896ed22550fb6a38f34/src/print.rs#L91C20-L91C20) so starships execution time is usually dominated by the biggest contributor. Looking at my timings in a git repo, thats usually git in the 20ms range:

```shell
[11:05:01] λ  starship timings

 Here are the timings of modules in your prompt (>=1ms or output):
 git_metrics    -  23ms  -   ""
 git_status     -  21ms  -   ""
 python         -  13ms  -   " v3.11.6 (venv:whatever-service-py3.11) "
 git_branch     -   3ms  -   "on  feature/som_branch "
 kubernetes     -   3ms  -   "󱃾 M3 in "
 git_commit     -   2ms  -   "(0e59472c) "
 aws            -   1ms  -   ""
 directory      -   1ms  -   "~/projects/whatever-service "
 time           -   1ms  -   "[11:17:59] "
 battery        -   1ms  -   ""
 pijul_channel  -   1ms  -   ""
 character      -  <1ms  -   "λ  "
 line_break     -  <1ms  -   "\n"
```

[the remark this answers is not in the document]
Comment (davidkna) [reply src/modules/direnv.rs:27]: It might be worth implementing the direnv allow checks ourselves, looking at the direnv config, checking `~/.local/share/direnv/allow/*` etc.
As a command that is always run, avoiding running `direnv status` seems worthwhile. But, I feel that as an initial implementation, the performance of this PR appears to be good enough for now.

[the remark this answers is not in the document]
Comment (Ajpantuso) [reply src/modules/direnv.rs:27]: Yes, I think with iteration we can improve this. My initial concern is just that direnv does not provide a public interface without running `direnv status` so coupling to the current implementation could make this module a maintenance burden. I am definitely not certain that the existing on-disk data formats are consistent across direnv versions today so could be tricky.

For the moment I'd prefer to get it in the hands of end users to see what user feedback dictates should be prioritized.

File: src/modules/direnv.rs [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+        std::fs::File::create(&rc_path)?.sync_all()?;
+
+        let renderer = ModuleRenderer::new("direnv")
+            .config(toml::toml! {
+                [direnv]
+                disabled = false
+            })
+            .path(dir.path())
+            .cmd(
+                "direnv status",
```
[the remark this answers is not in the document]
Comment (Ajpantuso) [reply src/modules/direnv.rs:164]: Yes, that would introduce additional logic for backwards compatibility with older versions, but is definitely a more stable source of information.

**Review by davidkna — APPROVED** [review]
LGTM

**Comment by andytom** [comment]
Thank you for your contribution @Ajpantuso and thank you for reviewing @davidkna.

**Comment by K900** [comment]
For some reason, my `direnv status` output has `Found RC allowed 0` instead of `false`, which breaks this :(

**Comment by davidkna** [comment]
@K900 #5657 is supposed to fix this I think.

**Comment by K900** [comment]
Thanks!

**Comment by youtux** [comment]
Hi, the plugin doesn't seem to work for me. I get the followign error when `cd`ing into a folder managed with direnv: 
```
[WARN] - (starship::modules::direnv): invalid allow status
```

Debug info:
```
$ pwd
/Users/youtux/Developer/pytest-factoryboy

$ cat .envrc
layout poetry

$ direnv status 
direnv exec path /opt/homebrew/bin/direnv
DIRENV_CONFIG /Users/youtux/.config/direnv
bash_path /opt/homebrew/bin/bash
disable_stdin false
warn_timeout 5s
whitelist.prefix []
whitelist.exact map[]
Loaded RC path /Users/youtux/Developer/pytest-factoryboy/.envrc
Loaded watch: ".envrc" - 2022-08-19T20:16:58+02:00
Loaded watch: "../../.local/share/direnv/allow/05157af877ed675cf491fd34d680cbe7fd7e3c3577f22dd0c3b3daa7a955a35f" - 2024-01-06T10:45:19+01:00
Loaded watch: "../../.local/share/direnv/deny/34cd9554aa44c1d0339de5a3ea1f7ca9fdef424dc3ff07a4edafb9f9819d3a7e" - 1970-01-01T01:00:00+01:00
Loaded RC allowed 0
Loaded RC allowPath /Users/youtux/.local/share/direnv/allow/05157af877ed675cf491fd34d680cbe7fd7e3c3577f22dd0c3b3daa7a955a35f
Found RC path /Users/youtux/Developer/pytest-factoryboy/.envrc
Found watch: ".envrc" - 2022-08-19T20:16:58+02:00
Found watch: "../../.local/share/direnv/allow/05157af877ed675cf491fd34d680cbe7fd7e3c3577f22dd0c3b3daa7a955a35f" - 2024-01-06T10:45:19+01:00
Found watch: "../../.local/share/direnv/deny/34cd9554aa44c1d0339de5a3ea1f7ca9fdef424dc3ff07a4edafb9f9819d3a7e" - 1970-01-01T01:00:00+01:00
Found RC allowed 0
Found RC allowPath /Users/youtux/.local/share/direnv/allow/05157af877ed675cf491fd34d680cbe7fd7e3c3577f22dd0c3b3daa7a955a35f

$ direnv status --json
{
  "config": {
    "ConfigDir": "/Users/youtux/.config/direnv",
    "SelfPath": "/opt/homebrew/bin/direnv"
  },
  "state": {
    "foundRC": {
      "allowed": 0,
      "path": "/Users/youtux/Developer/pytest-factoryboy/.envrc"
    },
    "loadedRC": {
      "allowed": 0,
      "path": "/Users/youtux/Developer/pytest-factoryboy/.envrc"
    }
  }
}

$ direnv --version
2.33.0

$ starship --version
starship 1.17.0
branch:
commit_hash:
build_time:2023-12-28 18:29:58 +00:00
build_env:rustc 1.74.0 (79e9716c9 2023-11-13) (Homebrew),
```

**Comment by axe312ger** [comment]
@youtux can confirm what he got :(

**Comment by glasser** [comment]
This module only triggers if you're in a directory with a `.envrc` file, not a child of such a directory.  (I know I can configure the `detect_*` options but none of them let you say "run in any directory where a parent has `.envrc`" or even "just run everywhere".) I think it would be nice if there were a way to trigger this based on "is `DIRENV_DIR` set" rather than based on files (as was suggested in a thread above). Would a PR be taken that supports this? Not sure how exactly to configure this, though honestly it kind of feels like it would be a reasonable default, and better than the standard `detect_*` thing.
